Develop

jussiemion opened the pull request.

ruscoder commented on src/services/call.ts:
```diff
@@ -108,6 +109,30 @@ export class CallService {
         return this.myInstance;
     }
 
+    public static async makeOutgoingCall(user: User, props: any) {
```
I don't like this code. Call service shouldn't have knowledge about component props. This service was architected as a reusable service between different projects. And also this service shouldn't have knowledge about Navigation.

You should just pass the method makeOutgoingCall from Call to Chat and handle isPending state in Chat component

ruscoder replied:
It isn't the only solution. Another solution is to create a service like fhir/saveFHIRResource, which accepts cursor to RemoteData.
But the first is simpler to implement.

Commit:
Create makeOutgoingCall method in Call service

Commit:
Ref #3 Add call button

Commit:
Extend .gitignore

Commit:
Rewrite calls with makeOutgoingCall from Call service

ruscoder commented on src/services/call.ts:
```diff
@@ -132,6 +134,33 @@ export class CallService {
+    ): Promise<RemoteDataResult<Voximplant['Call']>> {
+        cursor.set(loading);
+
+        try {
```
You don't need to use 'try' block twice. Just handle all errors in `catch` block

ruscoder commented on src/containers/ContactList/index.tsx:
```diff
@@ -110,19 +112,24 @@ export class Component extends React.Component<ComponentProps, {}> {
-                        passProps: { text: err.message },
-                    },
-                });
+            const callResponse = await CallService.makeOutgoingCall(
```
Don't use try here. Errors are handled in `makeOutgoingCall` service. Here you have a RemoteDataResult which can be a success or a failure. 

Just show an error in case of a failure.
And in all cases restore `isPending` flag.

ruscoder commented on src/containers/Chat/index.tsx:
```diff
@@ -151,13 +153,39 @@ export class Component extends React.Component<ComponentProps, {}> {
-                await Navigation.pop(this.props.componentId);
-            } finally {
-                this.props.tree.isPending.set(false);
+                const User = {
```
Change `makeOutgoingCall` service to accept username and displayName only. Don't construct a User instance, it is wrong because you cannot provide all User fields.

Commit:
Refactor Call service

- remove excess try ... catch construction
- split object with user data to username and displayName arguments

Commit:
Refactor logic of calls

- remove excess object with user data
- replace arguments in makeOutgoingCall methods
- rewrite try ... catch constructions with if ... else
- add missed isPending flag restoring

Commit:
Remove ramda and ramda-adjunct dependencies

Commit:
Refactor Call service and calls logic

- call.ts: put missed pushing result into cursor
- contactList/index.ts: simplify conditional construct
- Chat/index.ts: replace try ... catch costruction with conditional
construct
